### PR TITLE
feat: add parameter value origin field to parameters

### LINKED ISSUE #1454: New property to record additional info when control owners update parameter values
## Issue description / feature objectives
As a control owner, when updating parameter values in the yaml header in the markdown file, it is essential to record if the value is obtained from a specific corporate policy or decided by myself. Currently there is no property that can hold this value.  Hence this issue to add a new property. The type of the property is a free text field.

## Caveats / Assumptions
Unknown / None

## Completion Criteria
New property is added to each of the control markdown file.  When the markdown is regenerated from the profile, already saved data for this property isn't lost.

### PR DESCRIPTION
**Pull request by AleJo2995**
## Types of changes

- [ ] Hot fix (emergency fix and release)
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Documentation (change which affects the documentation site)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Release (`develop` -> `main`)

## Quality assurance (all should be covered).

- [x] My code follows the code style of this project.
- [x] Documentation for my change is up to date?
- [x] My PR meets testing requirements.
- [x] All new and existing tests passed.
- [x] All commits are signed-off.

## Summary

Closes #1454 

## Key links:

- [Sonar coverage](https://sonarcloud.io/dashboard?id=compliance-trestle)

# Before you merge

- Ensure it is a 'squash commit' if not a release.
- Ensure CI is currently passing
- Check sonar. If you are working for a fork a maintainer will reach out, if required.

### CODE REVIEW

File: trestle/oscal/common.py [hunk 1]
```diff
@@ -1626,6 +1626,7 @@ class Config:
     values: Optional[List[constr(regex=r'^\S(.*\S)?$')]] = Field(None)
     select: Optional[ParameterSelection] = None
     remarks: Optional[str] = None
+    param_value_origin: Optional[str] = None
```
Comment (vikas-agarwal76) [inline trestle/oscal/common.py:1629]: This is wrong. We are not adding a field directly to param origin in OSCAL. We are just adding a property. So OSCAL schema should not be changed.

File: trestle/oscal/profile.py [hunk 2]
```diff
@@ -106,6 +106,7 @@ class Config:
     guidelines: Optional[List[common.ParameterGuideline]] = Field(None)
     values: Optional[List[constr(regex=r'^\S(.*\S)?$')]] = Field(None)
     select: Optional[common.ParameterSelection] = None
+    param_value_origin: Optional[str] = Field(None, alias='parameter-value-origin')
```
Comment (vikas-agarwal76) [inline trestle/oscal/profile.py:110]: This is wrong. We are not changing the schema.

**Review by vikas-agarwal76 — CHANGES_REQUESTED** [review]
Need to discuss.

File: trestle/core/catalog/catalog_writer.py [hunk 3]
```diff
@@ -190,14 +192,21 @@ def _construct_set_parameters_dict(
             # adds display name, if no display name then do not add to dict
             if display_name != '' and display_name is not None:
                 new_dict[const.DISPLAY_NAME] = display_name
+            # adds param_Value_origin
+            if param_value_origin != '' and param_value_origin is not None:
+                new_dict[const.PARAM_VALUE_ORIGIN] = param_value_origin
+            else:
+                # puts it empty for user to fill it out mandatorily
+                new_dict[const.PARAM_VALUE_ORIGIN] = 'Added by Control Owner'
```
Comment (AleJo2995) [inline trestle/core/catalog/catalog_writer.py:200]: This value needs to be changed after Prashant confims which is the default value for the field. @vikas-agarwal76 . In the meantime I'm using a placeholder value

File: tests/trestle/core/commands/author/profile_test.py [hunk 4]
```diff
@@ -1300,3 +1300,63 @@ def test_profile_generate_assesment_objectives(tmp_trestle_dir: pathlib.Path, mo
...
+    # grabs first parameter in line and test out the value
+    assert profile.modify.set_parameters[0].props[1].value == 'coming from xyz corporate policy'
+
+    profile.modify.set_parameters[0].props[1].value = 'this is a change test'
+
+    ModelUtils.save_top_level_model(profile, tmp_trestle_dir, 'my_assembled_prof', FileContentType.JSON)
+
+    # convert resolved profile catalog to markdown then assemble it after adding an item to a control
+    # generate, edit, assemble
+    test_args = f'trestle author profile-generate -n {assembled_prof_name} -o {md_name} -rs NeededExtra --force-overwrite'.split(  # noqa E501
```
Comment (AleJo2995) [inline tests/trestle/core/commands/author/profile_test.py:1356]: Test includes overwrite of the profile at the end to proof that the markdown gets modified after a profile json is modified as well for param_value_origin prop

**Comment by AleJo2995** [comment]
@vikas-agarwal76 after profile assemble, the prop for param_value_origin will look like this:

image

File: trestle/core/catalog/catalog_writer.py (hunk 3 above)
Comment (vikas-agarwal76) [reply trestle/core/catalog/catalog_writer.py:200]: Is it required to populate the value for param value origin?

**Review by vikas-agarwal76 — COMMENTED** [review]
I didn't find any changes to profile code and Mardkown code for reading and writing param value origin. Where are those changes. Also if the param-value-origin was set in previous catalog or profile then how will you distinguish whether it wa set in this profile or pervious profile for including in profile->alters->set-params?

**Comment by AleJo2995** [comment]
@vikas-agarwal76 done. I have added the new test required for testing the inherited param-value-origin coming from another profile called: `test_param_value_origin_from_inherited_profile`

**Comment by AleJo2995** [comment]
@vikas-agarwal76 , @mrgadgil I have changed the default value for param-value-origin correctly. Could you please approve so I can merge? Thanks

**Review by mrgadgil — APPROVED** [review]

**Review by vikas-agarwal76 — APPROVED** [review]
LGTM for param-value-origin
@AleJo2995 Please open another issue and PR to fix the param-values issue that we discussed.

**Pull request merged**